beacon-chain/db/kv/state.go

```diff
@@ -99,30 +114,112 @@ func (s *Store) SaveStates(ctx context.Context, states []iface.ReadOnlyBeaconSta
+				if encodeErr != nil {
+					return encodeErr
+				}
+
+				// create the unique hash for that validator entry.
+				hash := hashutil.Hash(valBytes)
+				hashes = append(hashes, hash[:]...)
+
+				// note down the hash and the encoded validator entry
+				hashStr := hexutils.BytesToHex(hash[:])
+				validatorsEntries[hashStr] = valBytes
+			}
+
+			// zero out the validators List from the state bucket so that it is not stored as part of it.
+			pbState.Validators = nil
+
+			validatorKeys[i] = snappy.Encode(nil, hashes)
+		}
 		multipleEncs[i], err = encode(ctx, pbState)
 		if err != nil {
 			return err
 		}
 	}
 
-	return s.db.Update(func(tx *bolt.Tx) error {
-		bucket := tx.Bucket(stateBucket)
-		for i, rt := range blockRoots {
-			indicesByBucket := createStateIndicesFromStateSlot(ctx, states[i].Slot())
-			if err := updateValueForIndices(ctx, indicesByBucket, rt[:], tx); err != nil {
-				return errors.Wrap(err, "could not update DB indices")
+	if err := s.db.Update(func(tx *bolt.Tx) error {
+		if featureconfig.Get().EnableHistoricalSpaceRepresentation {
+			bucket := tx.Bucket(stateBucket)
+			valIdxBkt := tx.Bucket(blockRootValidatorHashesBucket)
+			for i, rt := range blockRoots {
+				indicesByBucket := createStateIndicesFromStateSlot(ctx, states[i].Slot())
+				if err := updateValueForIndices(ctx, indicesByBucket, rt[:], tx); err != nil {
+					return errors.Wrap(err, "could not update DB indices")
+				}
+				if err := bucket.Put(rt[:], multipleEncs[i]); err != nil {
+					return err
+				}
+				if err := valIdxBkt.Put(rt[:], validatorKeys[i]); err != nil {
+					return err
+				}
+			}
+
+			// store the validator entries separately to save space.
+			valBkt := tx.Bucket(stateValidatorsBucket)
+			for hashStr, validatorEntry := range validatorsEntries {
+				key := hexutils.HexToBytes(hashStr)
+				// if the entry is not in the cache and not in the DB,
+				// then insert it in the DB and add to the cache.
+				if _, ok := s.validatorEntryCache.Get(key); !ok {
+					validatorEntryCacheMiss.Inc()
+					if valEntry := valBkt.Get(key); valEntry == nil {
+						if putErr := valBkt.Put(key, validatorEntry); putErr != nil {
+							return putErr
+						}
+						s.validatorEntryCache.Set(key, validatorEntry, int64(len(validatorEntry)))
+					}
+				} else {
+					validatorEntryCacheHit.Inc()
+				}
 			}
-			if err := bucket.Put(rt[:], multipleEncs[i]); err != nil {
+			return nil
+		} else {
+			bucket := tx.Bucket(stateBucket)
+			for i, rt := range blockRoots {
+				indicesByBucket := createStateIndicesFromStateSlot(ctx, states[i].Slot())
+				if err := updateValueForIndices(ctx, indicesByBucket, rt[:], tx); err != nil {
+					return errors.Wrap(err, "could not update DB indices")
+				}
+				if err := bucket.Put(rt[:], multipleEncs[i]); err != nil {
+					return err
+				}
+			}
+			return nil
+
+		}
+	}); err != nil {
+		return err
+	}
+
+	if featureconfig.Get().EnableHistoricalSpaceRepresentation {
```

Comment:
What if someone uses the flag once, but next time they run Prysm they remove the flag. Meaning they have some states stored by using the flag and then next time the code reads from the DB without hitting this code branch if they removed the flag. Is it safer to check if we have performed the migration by reading the migration key from the DB? Once you enable historical representation, you should not be able to go back unless you migrate down. We should enforce that beyond simple feature flag conditionals

Reply:
Throwing error when the feature is disabled and migration is done.

Reply:
Very good point!

Reply:
Nice catch. Addressed.